### PR TITLE
ES/02 runner

### PR DESCRIPTION
**Pull request by samarjeet**
# ALCHEMI Toolkit Pull Request

EnhancedSampling runner (no exchange): walker identity stamping, force-step ordering, update() exactly-once, evaluate-only force priming, warm_start(); AdaptivePotentialMixin; HarmonicUmbrellaBias; LowerWall / UpperWall / FlatBottomRestraint; periodic_difference; unit tests; gallery example; user guide

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

-
-
-

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/enhanced_sampling/_runner.py [hunk 1]
```diff
@@ -0,0 +1,847 @@
...
+    compile_biases:
+        When ``True``, ``torch.compile`` each conservative bias's
+        ``energy()``.  Not ``evaluate()`` — that path calls
+        ``requires_grad_()``, which ``torch.compile`` cannot trace; see the
+        :class:`~nvalchemi.enhanced_sampling.ConservativeBias` docstring.
+    prime_after_update:
+        When ``True`` (default), re-evaluate biases and rewrite the batch's
+        total forces after an ``update()`` bumps a bias's state version, so
+        that anything reading ``batch.forces`` between steps sees the current
+        bias rather than the previous one.
```
Comment (dallasfoster) [inline nvalchemi/enhanced_sampling/_runner.py:125]: `BaseDynamics` already owns the stepping loop, so a second `run()` means every
future workflow has to decide which loop it belongs to, and the two will drift.

The runner's remaining jobs are smaller than they look. It stamps identity
fields onto the batch, fires things on a cadence, and writes a checkpoint. The
toolkit already has mechanisms for the first two — `register_bookkeeping_key`
(the same mechanism `status` and `system_id` use, so the fields survive refill
and graduation) and `Hook.frequency`. The third is row 8. Note that #176 has to
add `state_dict` to `BaseDynamics` anyway for the runner to work, which is
already most of what a strategy would need.

```python
# nvalchemi/dynamics/strategy.py — mirrors TrainingStrategy
class DynamicsStrategy(BaseModel):
    """Declarative recipe that builds and runs a configured BaseDynamics."""

    engine: type[BaseDynamics] = NVTLangevin
    engine_kwargs: dict[str, Any] = {}
    n_steps: int | None = None
    extra_hooks: list[Hook] = []          # runtime, not serialized

    def build_hooks(self) -> list[Hook]: ...
    def build(self, model) -> BaseDynamics: ...
    def run(self, batch, model, n_steps=None) -> Batch: ...
    def to_spec_dict(self) -> dict[str, Any]: ...

# nvalchemi/enhanced_sampling/strategy.py  (~80 lines, was 1414)
class EnhancedSampling(DynamicsStrategy):
    biases: dict[str, BiasPotential] = {}
    steps_per_epoch: int = 10_000

    def build_hooks(self) -> list[Hook]:
        return [WalkerIdentityHook(),
                *self.biases.values(),
                EpochCommitHook(frequency=self.steps_per_epoch)]

BaseDynamics.register_bookkeeping_key(
    "walker_id", lambda n, dev: torch.arange(n, device=dev).reshape(n, 1)
)
```

`TrainingStrategy` is the precedent worth copying: it is a Pydantic model that
configures a loop it does not own, and it serializes to a spec. Under this shape
`NEB`, `Relax` and `EOSScan` become sibling subclasses rather than sibling
runners.

**Review by dallasfoster — CHANGES_REQUESTED** [review]
